### PR TITLE
8 post a new comment to an existing article

### PR DESCRIPTION
**Pull request by TomMetcalf**

### CODE REVIEW

File: src/App.jsx [hunk 1]
```diff
@@ -5,19 +5,58 @@ import Nav from './components/Nav';
...
-        <Header />
-        <Nav />
-        <Routes>
-          <Route path="/" element={<ArticleList />}></Route>
-          <Route path="/topics" element={<TopicList />}></Route>
-          <Route path="/articles/:article_id" element={<Article />}></Route>
-        </Routes>
-      </>
+      <UserContext.Provider value={{user, setUser}}>
+        <main>
```
Comment (JVal-NC) [inline src/App.jsx]: Is main definitely needed here?

File: src/components/UserList.jsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+export default function UserList({ userList }) {
+  const [users, setUsers] = useState([]);
+  const [isLoading, setIsLoading] = useState();
+  const { setUser } = useContext(UserContext);
+  const [selectedUser, setSelectedUser] = useState();
+
+  useEffect(() => {
+    setIsLoading(true);
+    setUsers(userList.users);
```
Comment (JVal-NC) [inline src/components/UserList.jsx:13]: Can we put the set loading state to false inside a then block to ensure our asynchronous code runs at the correct.

Comment (JVal-NC) [inline src/components/UserList.jsx]: There's some duplication of state going on here I believe? We have userList as a state in the parent component but then also passed to the users state locally to this component. Is there areason for this as I suspect looking at how you've managed your context that there may be a problem you've attempted to solve with this?

Comment (JVal-NC) [reply src/components/UserList.jsx:13]: There's some duplication of state going on here I believe? We have userList as a state in the parent component but then also passed to the users state locally to this component. Is there a reason for this as it seems unnecessary to me?

File: src/App.jsx [hunk 3]
```diff
@@ -5,19 +5,58 @@ import Nav from './components/Nav';
...
 
 function App() {
+  const [userList, setUserList] = useState([]);
+  const [isLoading, setIsLoading] = useState();
+
+  const [user, setUser] = useState({
+    username: 'Logged Out',
+  });
+
+  useEffect(() => {
```
Comment (JVal-NC) [inline src/App.jsx]: This seems redundant, take a look at my UsersList comments.

File: src/components/CommentAdder.jsx [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    postComment(user.username, newComment, article_id)
+      .then((newCommentFromApi) => {
+        setNewComment('');
+        setComments((currComments) => {
+          return [...currComments, newCommentFromApi];
+        });
+      })
+      .catch((err) => {
```
Comment (JVal-NC) [inline src/components/CommentAdder.jsx:34]: Can we feed this back to the user somehow?

File: src/components/CommentAdder.jsx [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        console.log(
```
Comment (JVal-NC) [inline src/components/CommentAdder.jsx]: Can we feed this back to the user somehow without them needing to open the console?

File: src/components/CommentAdder.jsx [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+          `Failed to add comment for ${article_id}. Please try again.`
+        );
+      });
+  };
+
+  const handleTextareaFocus = () => {
+    setCommentAlert(false);
+  };
+
+  return (
```
Comment (JVal-NC) [inline src/components/CommentAdder.jsx:51]: Really great ammount of features in this so far Tom, great work.

Couple things to add are:
1. Preventing duplicate clicks of the post comment button from posting the same comment twice.
2. Some feedback for succesfull posts.

File: src/utils.js [hunk 7]
```diff
@@ -36,7 +36,33 @@ export function patchArticleVote(article_id, votes) {
...
+export function fetchUsers() {
+  return ncNewsApi
+    .get('/users')
+    .then((res) => {
+      return res.data;
+    })
+    .catch((err) => console.log(err));
+}
+
+export function postComment( username, newCommentText, article_id ) {
```
Comment (JVal-NC) [inline src/utils.js:56]: Nice work, could this file be renamed to api to reflect what the functions inside of it are for. utils may be deemed more generic.

**Review by JVal-NC — COMMENTED** [review]
Tom a few comments, but its a big ticket. Really good so far so keep it up!

**Review by Augs0 — CHANGES_REQUESTED** [review]
~~The users issues are preventing me from seeing your work on this ticket. Please fix them or use a hardcoded user. Take a look at the comments on your latest commit for some direction~~

Please view review comments

**Review by Augs0 — APPROVED** [review]
Well done, Tom, I can see now that's it loaded that you do have everything there for the ticket. I'll approve on this basis but strongly recommend you implement the changes I mentioned in your previous commit for a more robust app. At the moment, users is using two different approaches at the same time and it just happens to work i.e. mixing a prop being passed with a conditional render. Also consider if you can navigate a user back to the article they want to comment on after logging in for a slightly better UX.

**Pull request merged**